component/prometheus/source/api/api.go

```diff
@@ -0,0 +1,141 @@
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	"github.com/grafana/agent/component"
+	fnet "github.com/grafana/agent/component/common/net"
+	agentprom "github.com/grafana/agent/component/prometheus"
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "prometheus.source.api",
+		Args: Arguments{},
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+type Arguments struct {
+	Server    *fnet.ServerConfig   `river:",squash"`
```

Comment:
Up to you if you prefer to use the shared server implementation (probably for the best), just taking the opportunity to let you know about the HTTPComponent extension interface which can register a component's handler on the Agent's root HTTP server.

The handler is registered on a path unique to the component; so a call to `/component/{component_id}//api/v1/metrics/write` would reach the endpoint of the `prometheus.source.api "component_id"` instance.

https://github.com/grafana/agent/blob/d5b1d7dc1ee8b742f3a196e9c9c53181d85c27eb/component/component.go#L118-L127